File: src/java/foundgine-core/src/main/java/com/foundgine/core/execution/mutation/MutationDependencyLevels.java
package com.foundgine.core.execution.mutation;

import com.foundgine.core.semantic.planning.mutation.MutationDependency;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Set;

/**
 * Port of {@code Foundgine.Core.Execution.Mutation.MutationDependencyLevels}.
 *
 * <p>Computes execution levels directly from canonical mutation dependencies.
 * No correlation-shaped compatibility model is involved.
 */
public final class MutationDependencyLevels {

    private MutationDependencyLevels() {
    }

    public static List<List<Integer>> compute(int operationCount, Iterable<MutationDependency> dependencies) {
        if (operationCount < 0)
            throw new IllegalArgumentException("operationCount must not be negative.");
        Objects.requireNonNull(dependencies);

        List<MutationDependency> edges = new ArrayList<>();
        dependencies.forEach(edges::add);

        for (MutationDependency dependency : edges) {
            if (dependency.sourceOperationIndex() < 0 || dependency.sourceOperationIndex() >= operationCount
                    || dependency.targetOperationIndex() < 0 || dependency.targetOperationIndex() >= operationCount) {
                throw new IllegalStateException(
                        "Mutation dependency indexes are outside the execution graph: "
                                + dependency.sourceOperationIndex() + " -> " + dependency.targetOperationIndex() + ".");
            }

            if (dependency.sourceOperationIndex() == dependency.targetOperationIndex()) {
                throw new IllegalStateException(
                        "Mutation dependency cycle detected at operation " + dependency.sourceOperationIndex() + ".");
            }
        }

        if (operationCount == 0)
            return List.of();

        Map<Integer, Integer> incoming = new HashMap<>();
        Map<Integer, List<Integer>> outgoing = new HashMap<>();
        for (int i = 0; i < operationCount; i++) {
            incoming.put(i, 0);
            outgoing.put(i, new ArrayList<>());
        }

        for (MutationDependency dependency : edges) {
            List<Integer> targets = outgoing.get(dependency.sourceOperationIndex());
            if (!targets.contains(dependency.targetOperationIndex())) {
                targets.add(dependency.targetOperationIndex());
                incoming.merge(dependency.targetOperationIndex(), 1, Integer::sum);
            }
        }

        Set<Integer> remaining = new HashSet<>();
        for (int i = 0; i < operationCount; i++)
            remaining.add(i);

        List<List<Integer>> levels = new ArrayList<>();

        while (!remaining.isEmpty()) {
            List<Integer> level = remaining.stream()
                    .filter(i -> incoming.get(i) == 0)
                    .sorted()
                    .toList();

            if (level.isEmpty())
                throw new IllegalStateException("Mutation dependency graph contains a cycle.");

            levels.add(level);

            for (int node : level) {
                remaining.remove(node);
                for (int target : outgoing.get(node))
                    incoming.merge(target, -1, Integer::sum);
            }
        }

        return levels;
    }
}

File: src/csharp/Foundgine.Core/Execution/Mutation/MutationDependencyLevels.cs
using Foundgine.Core.Semantic.Planning.Mutation;

namespace Foundgine.Core.Execution.Mutation;

/// <summary>
/// Computes execution levels directly from canonical mutation dependencies.
/// No correlation-shaped compatibility model is involved.
/// </summary>
public static class MutationDependencyLevels
{
    public static IReadOnlyList<IReadOnlyList<int>> Compute(
        int operationCount,
        IEnumerable<MutationDependency> dependencies)
    {
        ArgumentOutOfRangeException.ThrowIfNegative(operationCount);
        ArgumentNullException.ThrowIfNull(dependencies);

        var edges = dependencies.ToArray();
        foreach (var dependency in edges)
        {
            if (dependency.SourceOperationIndex < 0 ||
                dependency.SourceOperationIndex >= operationCount ||
                dependency.TargetOperationIndex < 0 ||
                dependency.TargetOperationIndex >= operationCount)
                throw new InvalidOperationException(
                    $"Mutation dependency indexes are outside the execution graph: " +
                    $"{dependency.SourceOperationIndex} -> {dependency.TargetOperationIndex}.");

            if (dependency.SourceOperationIndex == dependency.TargetOperationIndex)
                throw new InvalidOperationException(
                    $"Mutation dependency cycle detected at operation {dependency.SourceOperationIndex}.");
        }

        if (operationCount == 0)
            return Array.Empty<IReadOnlyList<int>>();

        var incoming = Enumerable.Range(0, operationCount)
            .ToDictionary(i => i, _ => 0);
        var outgoing = Enumerable.Range(0, operationCount)
            .ToDictionary(i => i, _ => new List<int>());

        foreach (var dependency in edges)
        {
            if (!outgoing[dependency.SourceOperationIndex].Contains(dependency.TargetOperationIndex))
            {
                outgoing[dependency.SourceOperationIndex].Add(dependency.TargetOperationIndex);
                incoming[dependency.TargetOperationIndex]++;
            }
        }

        var remaining = Enumerable.Range(0, operationCount).ToHashSet();
        var levels = new List<IReadOnlyList<int>>();

        while (remaining.Count > 0)
        {
            var level = remaining
                .Where(i => incoming[i] == 0)
                .OrderBy(i => i)
                .ToArray();

            if (level.Length == 0)
                throw new InvalidOperationException(
                    "Mutation dependency graph contains a cycle.");

            levels.Add(level);

            foreach (var node in level)
            {
                remaining.Remove(node);
                foreach (var target in outgoing[node])
                    incoming[target]--;
            }
        }

        return levels;
    }
}
